core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BpmBillingProjectLifecycle.scala

```diff
@@ -97,26 +104,106 @@ class BpmBillingProjectLifecycle(
-                                                   landingZoneResponse.getLandingZoneId
-        )
+      (for {
+        _ <- billingRepository.updateLandingZoneId(createProjectRequest.projectName, landingZoneId.get)
         _ <- resourceMonitorRecordDao.create(
           UUID.fromString(landingZoneResponse.getJobReport.getId),
           JobType.AzureLandingZoneResult,
-          projectName
+          projectName.value
         )
         _ <- billingRepository.setBillingProfileId(createProjectRequest.projectName, profileModel.getId)
-      } yield CreationStatuses.CreatingLandingZone
+      } yield CreationStatuses.CreatingLandingZone).recoverWith { case t: Throwable =>
+        for {
+          _ <- cleanupLandingZone(landingZoneId, projectName, ctx)
+          _ <- cleanupBillingProfile(profileModelId, projectName, ctx)
+        } yield throw t
+      }
     } catch {
-      case exception: Exception => Future.failed(exception)
+      case t: Throwable =>
+        for {
+          _ <- cleanupLandingZone(landingZoneId, projectName, ctx)
+          _ <- cleanupBillingProfile(profileModelId, projectName, ctx)
+        } yield throw t
     }
   }
 
+  private def cleanupLandingZone(landingZoneId: Option[UUID],
+                                 projectName: RawlsBillingProjectName,
+                                 ctx: RawlsRequestContext
+  ): Future[Unit] =
+    landingZoneId match {
+      case Some(id) =>
+        val landingZoneResponse = Try(blocking {
+          // Note that this actually just starts a landing zone deletion job (and thus returns quickly).
+          // We are not attempting to ensure that the landing zone deletion completes successfully.
+          workspaceManagerDAO.deleteLandingZone(id, ctx)
+        })
+
+        landingZoneResponse match {
+          case Success(response) =>
+            Option(response) match {
+              case Some(result) =>
+                Option(result.getErrorReport) match {
+                  case Some(errorReport) =>
+                    logger.warn(
+                      s"Unable to delete landing zone with ID ${id} for BPM-backed billing project ${projectName.value}: ${errorReport.getMessage}."
+                    )
+                  case None =>
+                    logger.info(
+                      s"Initiated deletion of landing zone ${landingZoneId} for BPM-backed billing project ${projectName.value}."
+                    )
+                }
+              case None =>
+                logger.warn(
+                  s"Unable to delete landing zone with ID ${id} for BPM-backed billing project ${projectName.value}, no landingZoneResponse from WSM."
+                )
+            }
+          case Failure(e) =>
+            logger.warn(
+              s"Unable to delete landing zone with ID ${id} for BPM-backed billing project ${projectName.value}.",
+              e
+            )
+        }
+        Future.successful()
+      case None =>
```

Comment:
This case will only be hit during cleanup that happens due to a failed creation call (depending on where the error is hit, it could be expected that no landing zone exists). In the delete case we log it as a warning because we expect the landing zone to exist in that case.

Reply:
You can greatly simplify this function and `cleanupBillingProfile` by not having to take `Optional` (see above). Here's an example of how you could implement this. I'd be happy to work through the other one with you if you're interested. 
```scala
  private def cleanupLandingZone(landingZoneId: UUID,
                                 projectName: RawlsBillingProjectName,
                                 ctx: RawlsRequestContext
  ): Future[Unit] =
    (for {
      // Note that this actually just starts a landing zone deletion job (and thus returns quickly).
      // We are not attempting to ensure that the landing zone deletion completes successfully.
      landingZoneResponse <- Future(blocking {
        val response = workspaceManagerDAO.deleteLandingZone(landingZoneId, ctx)
        logger.info(
          s"Initiated deletion of landing zone $landingZoneId for BPM-backed billing project ${projectName.value}."
        )
        response
      })

     // Option.map (or traverse) is your friend when it comes to null-checking and throwing
      _ = Option(landingZoneResponse.getErrorReport).map { errorReport =>
        logger.warn(
          s"Unable to delete landing zone with ID $landingZoneId for BPM-backed " +
            s"billing project ${projectName.value}: ${errorReport.getMessage}."
        )
      }
    } yield ()).recover { case t: Throwable =>
      logger.warn(
        s"Unable to delete landing zone with ID $landingZoneId for BPM-backed billing project ${projectName.value}.",
        t
      )
    }
```